tempesta_fw/ss_skb.c

```diff
@@ -634,15 +634,17 @@ __split_pgfrag_del(struct sk_buff *skb_head, struct sk_buff *skb, int i, int off
 	if (likely(!off)) {
 		frag->page_offset += len;
 		skb_frag_size_sub(frag, len);
-		ss_skb_adjust_data_len(skb, -len);
+		skb->len -= len;
+		skb->data_len -= len;
```

Comment:
Not quite clear why we do not adjust `skb->truesize` here during removal the part of fragment, but in case of the full fragment deletion (in previous block, above) - we adjust it.
E.g. in another but similar case - in `ss_skb_split()` - during splitting the original skb  - we adjust `skb->truesize`, including the cases when splitting boundary located in the middle of the fragment of original skb: this can be seen in `skb_split() -> skb_split_no_header()` and we set `skb->truesize = skb->len - len` in `ss_skb_split()`.

Reply:
The reason for the change is that `skb->truesize` must account all the memory used by the skb. The accounted memory is required to for socket memory accounting: how much memory do we use by all skbs in the socket. If we call `skb_frag_size_sub()`, then we just move the grament pointer and don't free any memory, so the allocated memory for the skb remains the same.

Regarding `skb_split() -> skb_split_no_header()` I'd say that the kernel is very inconsistent in accounting `skb->truesize`. If you print `skb->len` and `skb->truesize` for some benchmark, then you can see difference 10-30KB for 40-60KB skbs due to clones and the framgents playing. See for example `skb_gro_receive()` for `skb->head_frag != NULL)` - the skb head is stolen to `head->frags`, `head` truesize is incremented and the data will be processed as its frag, but `skb` continues to account the head data in `truesize`. See also `pskb_expand_head()` which not always update truesize.

If you run a test against vanilla kernel with a small patch:
```
root@debian:~/linux-4.14.32-tfw# git log
commit 07d7b6fe5a8973b79b4e1bff2d9da889e3fd4379
Author: Alexander K <ak@natsys-lab.com>
Date:   Thu Apr 5 16:28:25 2018 +0300

    Vanilla Linux 4.14.32
root@debian:~/linux-4.14.32-tfw# git diff
diff --git a/net/ipv4/tcp_ipv4.c b/net/ipv4/tcp_ipv4.c
index cab4b935..87b7b9c1 100644
--- a/net/ipv4/tcp_ipv4.c
+++ b/net/ipv4/tcp_ipv4.c
@@ -1638,6 +1638,9 @@ int tcp_v4_rcv(struct sk_buff *skb)
        if (!pskb_may_pull(skb, sizeof(struct tcphdr)))
                goto discard_it;
 
+       if (skb->truesize > skb->len +2048)
+               pr_err("large truesize=%u > len=%u+2k\n", skb->truesize, skb->len);
+
        th = (const struct tcphdr *)skb->data;
 
        if (unlikely(th->doff < sizeof(struct tcphdr) / 4))
```
And run a benchmark envolvin heavy GRO, then you'll see truesizes much large than len. This is not the most large difference which I saw:
```
TCP: large truesize=19456 > len=14512+2k
```